### ISSUE TITLE
Can't connect to steam

### ISSUE THREAD

**Issue by xavierlaporte** [SteamRE/SteamKit#236]
Hi, I have been trying start up a steam bot with your sample programs

but only SteamClient.DisconnectedCallback callback triggers
so no connection to steam can be made.

thx for the help.

code:
<code>using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using SteamKit2;

namespace Sample1_Logon
{
    class Program
    {
        static SteamClient steamClient;
        static CallbackManager manager;

```
    static SteamUser steamUser;

    static bool isRunning;

    static string user, pass;


    static void Main(string[] args)
    {
        //if (args.Length < 2)
       // {
       //     Console.WriteLine("Sample1: No username and password specified!");
       //     return;
       // }

        // save our logon details
        user = "xxxxxx";
        pass = "xxxxxx";

        // create our steamclient instance
        steamClient = new SteamClient();
        // create the callback manager which will route callbacks to function calls
        manager = new CallbackManager(steamClient);

        // get the steamuser handler, which is used for logging on after successfully connecting
        steamUser = steamClient.GetHandler<SteamUser>();

        // register a few callbacks we're interested in
        // these are registered upon creation to a callback manager, which will then route the callbacks
        // to the functions specified
        manager.Subscribe<SteamClient.ConnectedCallback>(OnConnected);
        manager.Subscribe<SteamClient.DisconnectedCallback>(OnDisconnected);

        manager.Subscribe<SteamUser.LoggedOnCallback>(OnLoggedOn);
        manager.Subscribe<SteamUser.LoggedOffCallback>(OnLoggedOff);

        isRunning = true;

        Console.WriteLine("Connecting to Steam...");

        // initiate the connection
        steamClient.Connect();

        // create our callback handling loop
        while (isRunning)
        {
            // in order for the callbacks to get routed, they need to be handled by the manager
            manager.RunWaitCallbacks(TimeSpan.FromSeconds(1));
        }
    }

    static void OnConnected(SteamClient.ConnectedCallback callback)
    {
        Console.WriteLine("connectedCallBack");
        if (callback.Result != EResult.OK)
        {
            Console.WriteLine("Unable to connect to Steam: {0}", callback.Result);

            isRunning = false;
            return;
        }

        Console.WriteLine("Connected to Steam! Logging in '{0}'...", user);

        steamUser.LogOn(new SteamUser.LogOnDetails
        {
            Username = user,
            Password = pass,
        });
    }

    static void OnDisconnected(SteamClient.DisconnectedCallback callback)
    {
        Console.WriteLine("Disconnected from Steam");
        Console.ReadKey();
        isRunning = false;
    }

    static void OnLoggedOn(SteamUser.LoggedOnCallback callback)
    {
        if (callback.Result != EResult.OK)
        {
            if (callback.Result == EResult.AccountLogonDenied)
            {
                // if we recieve AccountLogonDenied or one of it's flavors (AccountLogonDeniedNoMailSent, etc)
                // then the account we're logging into is SteamGuard protected
                // see sample 5 for how SteamGuard can be handled

                Console.WriteLine("Unable to logon to Steam: This account is SteamGuard protected.");

                isRunning = false;
                return;
            }

            Console.WriteLine("Unable to logon to Steam: {0} / {1}", callback.Result, callback.ExtendedResult);

            isRunning = false;
            return;
        }

        Console.WriteLine("Successfully logged on!");

        // at this point, we'd be able to perform actions on Steam

        // for this sample we'll just log off
        steamUser.LogOff();
    }

    static void OnLoggedOff(SteamUser.LoggedOffCallback callback)
    {
        Console.WriteLine("Logged off of Steam: {0}", callback.Result);
    }
}
```

}</code>

**Comment by VoiDeD** [comment]
There's two issues at play here:

1. The first sample doesn't handle connection retry logic. I don't believe that any of our samples do, but essentially after handling `OnDisconnected`, you'd want to delay for a while (maybe 10 seconds or so), and then call `SteamClient.Connect` again to retry the connection.
2. SteamKit's internal list of Steam servers is getting outdated and contains some dead/moved servers.

When you call `SteamClient.Connect` without specifying an endpoint, SteamKit will randomly attempt to select the best server candidate from its internal list of servers. The list also maintains a "score" value that's tied to whether or not the server is connectable, and the random selection is skewed to try to select the server with the best score.

So with additional retry logic, SteamKit will eventually connect to a valid server (assuming that Steam isn't experiencing downtime).

We have [some sample code](https://github.com/SteamRE/SteamKit/blob/master/Samples/7.ServerList/Program.cs) that queries Steam to update SteamKit's internal list of servers (and it also handles caching this list to disk so that you don't have to hit Steam's web API every time to update the list), so the addition of that to your code should be all you need.

**Comment by JustArchi** [comment]
TL;DR - Add ```SteamDirectory.Initialize.Wait();``` before attempting to connect.

**Comment by VoiDeD** [comment]
I suppose the elegant science of "don't hit Steam's web API on every connection attempt" are lost on some.

**Comment by JustArchi** [comment]
It has to be done only once on program startup. I didn't suggest to do that before every attempt of connecting. Just before first one.

List should be up-to-date anyway, so saving it serves little to no purpose, as in long-run it will lead to the same situation as SK2 currently has.

**Comment by xavierlaporte** [comment]
Thanks JustArchi did the job

**Comment by Netshroud** [comment]
Once you've made the initial connection, Steam will give you a list of servers to use. You should prefer that over reloading the WebAPI every time - see the sample mentioned above for a sample implementation of this.

**Comment by KenBeanNet** [comment]
Seemed to also fix my issue.  

Also noticing a bit of issues after the connection is made, when attempt login I never get a response.  Ive tried many accounts locally and it works.  So its not an IP or account ban.  I run 10 bots and they all log and log out frequently.  When it occurs, it happens to all of them.  Each on different networks.  Its like valve is going down, but it doesnt fully go down as I can make a Steam Connection just cant login.

This only started on 1.7.0.  Anyone else have this?

**Comment by JustArchi** [comment]
@Netshroud 
What if we save that list, and then never again launch our project in half a year? Yes, I know it's kind of impossible, but our previously saved list will be outdated, hence I prefer to ask (once) on SK2-based project startup for list of CMs, then attempt to connect, instead of saving even up-to-date list in some kind of persistent storage.

**Comment by Netshroud** [comment]
It's a bit of a delicate situation, because we would need to detect "we have no available CMs" versus "someone unplugged our network cable".

My initial thought is to add an event/callback-type thing that fires when we run out of "good" servers. You could then choose to reload the list from WebAPI or similar.

**Comment by JustArchi** [comment]
I don't know, personally I'd just remove hardcoded list and add ```SteamDirectory.Initialize().Wait();``` inside SK2 initialization, because that request has to be done only once per program startup, and it doesn't cost much compared to whole activity that SK2 performs, it's just one simple web api request.

Sure, you can add option to use loaded server list instead, when creating ```SteamClient``` object or similar, so programmers can optionally save it and reuse from the HDD, but by default I'd just ask web api for that if not provided.

But that's just my 3 cents, I'm not here to decide.

**Comment by KenBeanNet** [comment]
I moved back down to 1.6.5 and im flying again.  Not sure what the difference is, but seems to be better overall with connection/staying connected.

**Comment by azuisleet** [comment]
There don't appear to be any significant changes in networking between 1.6.5 and 1.7.0 beyond support for the new channel encryption protocol. Assuming that the handshake is successful, I doubt the connection is breaking down. It seems more likely that 1.6.5 is more likely to pick a particular boostrap server that is alive.

If there are issues with the networking itself, posting output from the DebugLog would go a long way in diagnosing any issues.

**Comment by JustArchi** [comment]
But I must admit that I also noticed similar issues, I have no idea what is the root issue though.

**Comment by KenBeanNet** [comment]
Its not even comparable to me.  All I am doing is grabbing a replay salt in DoTa.  Getting a connection takes like 8 retries on 1.7.0.  1.6.5 its almost instant everytime.

**Comment by TimotheeJeannin** [comment]
I also have this problem with the latest version. It takes several retries to get a connection.
A filtered version of my logs looks like:

    2016-02-10 12:14:02,712 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.201:27017
    2016-02-10 12:14:12,698 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.201:27018
    2016-02-10 12:14:22,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.201:27019
    2016-02-10 12:14:32,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.201:27020
    2016-02-10 12:14:42,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.202:27017
    2016-02-10 12:14:52,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.202:27018
    2016-02-10 12:15:02,698 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.202:27019
    2016-02-10 12:15:12,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.203:27017
    2016-02-10 12:15:22,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.203:27018
    2016-02-10 12:15:32,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.203:27019
    2016-02-10 12:15:42,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.204:27017
    2016-02-10 12:15:52,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.204:27018
    2016-02-10 12:16:02,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.204:27019
    2016-02-10 12:16:12,699 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.205:27017
    2016-02-10 12:16:22,694 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.205:27018
    2016-02-10 12:16:32,695 DEBUG - TcpConnection: Timed out while connecting to 208.64.200.205:27019
    2016-02-10 12:16:37,803 DEBUG - TcpConnection: Connected to 208.78.164.9:27017

**Comment by TimotheeJeannin** [comment]
When I use version `1.6.4` the first server it tries to connect to is `208.78.164.14:27017` and a connection is established on the first attempt.

**Comment by yaakov-h** [comment]
Closing as this is now handled internally within SteamKit.
